llvm/include/llvm/BitstreamCAS/BitstreamCASWriter.h

```diff
@@ -0,0 +1,572 @@
+
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Bitstream/BitCodes.h"
+#include "llvm/BitstreamCAS//BitstreamCASNode.h"
+#include "llvm/CAS/CASID.h"
+#include "llvm/CAS/CASReference.h"
+#include "llvm/CAS/ObjectStore.h"
+#include "llvm/Support/Casting.h"
+#include "llvm/Support/Endian.h"
+#include "llvm/Support/Error.h"
+#include "llvm/Support/MathExtras.h"
+#include "llvm/Support/raw_ostream.h"
+#include <algorithm>
+#include <optional>
+#include <string>
+#include <utility>
+#include <vector>
+
+namespace llvm {
+
+class BitstreamCASWriter {
+  /// CAS - The object store that the bitstream is written to.
+  std::shared_ptr<cas::ObjectStore> &CAS;
+
+  /// NodeStack - This helps in tracking the parent node of the current node.
+  std::stack<cas::BitstreamCASNode> NodeStack;
+
+  /// CurBit - Always between 0 and 31 inclusive, specifies the next bit to use.
+  unsigned CurBit = 0;
+
+  /// CurValue - The current value. Only bits < CurBit are valid.
+  uint32_t CurValue = 0;
+
+  /// CurCodeSize - This is the declared size of code values used for the
+  /// current block, in bits.
+  unsigned CurCodeSize = 2;
+
+  /// BlockInfoCurBID - When emitting a BLOCKINFO_BLOCK, this is the currently
+  /// selected BLOCK ID.
+  unsigned BlockInfoCurBID = 0;
+
+  /// CurAbbrevs - Abbrevs installed at in this block.
+  std::vector<std::shared_ptr<BitCodeAbbrev>> CurAbbrevs;
+
+  // Support for retrieving a section of the output, for purposes such as
+  // checksumming.
+  std::optional<size_t> BlockFlushingStartPos;
+
+  struct Block {
+    unsigned PrevCodeSize;
+    size_t StartSizeWord;
+    std::vector<std::shared_ptr<BitCodeAbbrev>> PrevAbbrevs;
+    Block(unsigned PCS, size_t SSW) : PrevCodeSize(PCS), StartSizeWord(SSW) {}
+  };
+
+  /// BlockScope - This tracks the current blocks that we have entered.
+  std::vector<Block> BlockScope;
+
+  /// BlockInfo - This contains information emitted to BLOCKINFO_BLOCK blocks.
+  /// These describe abbreviations that all blocks of the specified ID inherit.
+  struct BlockInfo {
+    unsigned BlockID;
+    std::vector<std::shared_ptr<BitCodeAbbrev>> Abbrevs;
+  };
+  std::vector<BlockInfo> BlockInfoRecords;
+
+  void WriteWord(unsigned Value) {
+    Value =
+        support::endian::byte_swap<uint32_t, llvm::endianness::little>(Value);
+    NodeStack.top().Data.append(reinterpret_cast<const char *>(&Value),
+                                reinterpret_cast<const char *>(&Value + 1));
+  }
+
+  size_t GetWordIndex() const {
+    size_t Offset = NodeStack.top().Data.size();
+    assert((Offset & 3) == 0 && "Not 32-bit aligned");
+    return Offset / 4;
+  }
+
+public:
+  /// Create a BitstreamCASWriter that writes to Content Addressable Storage.
+  /// \p CAS is the object store that the bitstream content is written to.
+  BitstreamCASWriter(std::shared_ptr<cas::ObjectStore> &CAS)
+      : CAS(CAS), NodeStack({cas::BitstreamCASNode()}) {}
```

Comment:
Nit: Default initializer for NodeStack is good enough. No need to explicitly init NodeStack

Reply:
Sorry, I'm not sure if I understand the suggestion correctly. The reason I am using the initialiser list in constructor and starting the NodeStack with a single node is to represent the RootNode of the entire bitstream file. Do you suggest adding the initial root node in the NodeStack somewhere else instead of constructor?

Reply:
It is very unclear what is happening here. I will either lazy init the first element when needed, or just explicitly initialize first element in the constructor body.

Reply:
Yeah, I agree. I have initialized first element in constructor body and added a comment there.